Add ForkContext

trowski opened the pull request:
This adds a `Context` implementation which uses `pcntl_fork` to create another process. This is a strategy which we abandoned some time ago due to the issues with copying parent context into a child, however there are particular case where that behavior can be advantageous.

This context is NOT added to `DefaultContextFactory`, and in fact has warnings against its use in the class docblock.

@danog Can you please check if this works for you in Psalm.

trowski committed:
Add ForkContext

danog commented:
This does not work in psalm, because `runContext` is queueing a callback onto the event loop, which causes an immediate `The event loop is already running` error when hitting EventLoop::run(); unwrapping the callback fixes the issue.

danog commented:
Additionally, I added some extra error handling logic to handle segfaults in child processes (commonly caused by JIT usage in psalm), would love to see that in this impl as well: https://github.com/vimeo/psalm/blob/fc437245b15a7220de107bc191a3559fdbf94595/src/Psalm/Internal/Fork/ForkContext.php#L61

danog commented:
`checkExit` should probably also be executed within the constructor in a finally block as well, as https://github.com/vimeo/psalm/issues/11310 is caused by a crash quite possibly before the initial IPC handshake is completed.

trowski committed:
Updates

trowski commented:
> This does not work in psalm, because runContext is queueing a callback onto the event loop, which causes an immediate The event loop is already running error when hitting EventLoop::run(); unwrapping the callback fixes the issue.

My local test was running from {main}, so I missed that. Unfortunately running this new context through the usual test suite crashes and burns, of course because we're forking the test process.

I removed the event loop queuing from `runContext` and updated `ForkContext` with the recent changes you made. Give it another try when you get a chance.

trowski committed:
Fix

kelunik commented:
What is the use case you mentioned?

trowski commented:
> What is the use case you mentioned?

@danog Can you please elaborate on how you're using this in Psalm?

danog commented:
I'm using this in psalm to avoid serializing and sending tens of megabytes of data structures containing information about the codebase post-scanning, when starting the analysis phase.

danog committed:
Fixup tests

danog committed:
Fixup

danog commented:
Ping @trowski  :)

trowski committed:
Merge branch '2.x' into fork-context

danog committed:
Fork context (#214)

* Skip another test

* cs-fix

kelunik commented on src/Context/ForkContext.php:
```diff
@@ -0,0 +1,170 @@
+    public static function isSupported(): bool
+    {
+        return \function_exists('pcntl_fork')
+            && !EventLoop::getDriver() instanceof UvDriver;
```
Docs say it's not compatible with extension based event loops, but we only check for UV here.

danog replied:
Yeah, this needs fixing, @trowski mind bumping

kelunik commented on src/Context/ForkContext.php:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        // Child
+        \define("AMP_CONTEXT", "parallel");
```
Is this the correct value or due to copy and paste?

trowski replied:
Oops, yes, forgot this constant isn't the library name, but the context name.

kelunik commented on src/Context/Internal/functions.php:
```diff
@@ -6,79 +6,81 @@
+    Serializer $serializer = new NativeSerializer(),
+): void {
+    /** @noinspection PhpUnusedLocalVariableInspection */
+    $argc = \count($argv);
```
Why did this code change in this PR?

trowski replied:
1) The call to `EventLoop::run()` was removed because the event loop is already running when called from `ForkContext`
2) Added the ability to pass a particular serializer instance, also used in `ForkContext`.

kelunik commented on test/Context/ForkContextTest.php:
```diff
@@ -0,0 +1,37 @@
+
+    public function testImmediateJoin(): void
+    {
+        // tmp
```
What about these?

trowski replied:
@danog This was copied from your code, can you explain?

danog replied:
Don't have a clear explanation for these specific failures yet, but the current state is better than the initial state of the PR which completely disabled tests

trowski committed:
Fix copy/pasta mistake

trowski committed:
Look for StreamSelectLoop

trowski committed:
Unused import

trowski committed:
Merge branch '2.x' into fork-context

trowski committed:
Fix loop type check

trowski committed:
Merge branch 'refs/heads/2.x' into fork-context

# Conflicts:
#	src/Context/Internal/functions.php

trowski committed:
Style fix after merge

trowski committed:
Close IPC channel immediately after callback finishes executing

trowski committed:
Remove commented-out tests

trowski committed:
A couple notes

trowski committed:
Distinguish between send and receive

trowski committed:
Check for extensions

trowski committed:
Don't call close in join

trowski committed:
Change variable name to match interface suffix
